warehouse/query-core/src/main/java/datawave/query/jexl/functions/TermFrequencyList.java

```diff
@@ -81,17 +82,31 @@ public String toString() {
     }
     
     /**
-     * Get an event id from the term frequency key
+     * Get an event id from the term frequency key. Assumes the key is valid, i.e., not missing a column qualifier and column qualifier is properly formatted.
+     *
+     * Transform a TF key like 'shard:tf:datatype\0uid\0value\0FIELD' into 'shard\0datatype\0uid'
      * 
      * @param key
+     *            a TermFrequency key
      */
     public static String getEventId(Key key) {
         StringBuilder eventId = new StringBuilder();
         eventId.append(key.getRow()).append('\0');
-        String cq = key.getColumnQualifier().toString();
-        int index = cq.indexOf('\0');
-        index = cq.indexOf('\0', index + 1);
-        eventId.append(cq, 0, index);
+        
+        int index = 0;
+        ByteSequence backing = key.getColumnQualifierData();
+        for (int i = backing.offset(); i < backing.length(); i++) {
+            if (backing.byteAt(i) == '\u0000') {
+                if (index == 0) {
+                    index = i; // first null byte
+                } else {
+                    index = i; // second null byte
+                    break;
+                }
+            }
+        }
+        
+        eventId.append(new String(backing.subSequence(0, index).toArray()));
```

Comment:
I think you can avoid the array copy implicit in the `ArrayByteSequence.subSequence()/toArray()` call if you just use the `String(byte[], int, int)` constructor on the original ByteSequence's backing here instead.

Reply:
Good catch, making this change